library/core/resource-loader/src/main/java/org/quiltmc/qsl/resource/loader/mixin/KeyedResourceReloaderMixin.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2021 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.resource.loader.mixin;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Locale;
+
+import org.spongepowered.asm.mixin.Mixin;
+
+import net.minecraft.loot.LootManager;
+import net.minecraft.recipe.RecipeManager;
+import net.minecraft.server.ServerAdvancementLoader;
+import net.minecraft.server.function.FunctionLoader;
+import net.minecraft.tag.TagManagerLoader;
+import net.minecraft.util.Identifier;
+
+import org.quiltmc.qsl.resource.loader.api.reloader.IdentifiableResourceReloader;
+import org.quiltmc.qsl.resource.loader.api.reloader.ResourceReloaderKeys;
+
+@Mixin({
+		RecipeManager.class, ServerAdvancementLoader.class, FunctionLoader.class,
+		LootManager.class, TagManagerLoader.class
+})
+public abstract class KeyedResourceReloaderMixin implements IdentifiableResourceReloader {
+	private Identifier quilt$id;
+	private Collection<Identifier> quilt$dependencies;
+
+	@Override
+	@SuppressWarnings({"ConstantConditions", "RedundantCast"})
+	public Identifier getQuiltId() {
+		if (this.quilt$id == null) {
+			Object self = this;
+
+			if (self instanceof RecipeManager) {
+				this.quilt$id = ResourceReloaderKeys.RECIPES;
+			} else if (self instanceof ServerAdvancementLoader) {
+				this.quilt$id = ResourceReloaderKeys.ADVANCEMENTS;
+			} else if (self instanceof FunctionLoader) {
+				this.quilt$id = ResourceReloaderKeys.FUNCTIONS;
+			} else if (self instanceof LootManager) {
+				this.quilt$id = ResourceReloaderKeys.LOOT_TABLES;
+			} else if (self instanceof TagManagerLoader) {
+				this.quilt$id = ResourceReloaderKeys.TAGS;
+			} else {
+				this.quilt$id = new Identifier("minecraft",
```

Comment:
Shouldn't this throw instead?

Reply:
Unsure, that's inheriting old FAPI code.

Reply:
This actually makes more sense in the client version of the mixin as some of them fall into the else block.

Reply:
I'd probably add a message guarded by the development environment noting the class if we don't have it covered by the statement.

Reply:
Meh, from the look of the old code those classes having for fallback this is pretty much intended.